sync15-adapter/src/token.rs

```diff
@@ -130,4 +148,84 @@ impl TokenserverClient {
 
         Ok(Authorization(format!("Hawk {}", header)))
     }
+
+    pub fn api_endpoint(&self) -> String {
+        self.api_endpoint.clone()
+    }
+}
+
+
+/// The TokenserverClient - long lived and fetches tokens on demand (eg, when
+/// first needed, or when an existing one expires.)
+#[derive(Debug)]
+pub struct TokenserverClient {
+    // The stuff needed to fetch a token.
+    base_url: String,
+    access_token: String,
+    key_id: String,
+    // The token (and derived info) that we fetch and hand back to consumers.
+    // While this seems odd, we use an Rc to try and ensure consumers can't
+    // hold on to it - we'll panic if we try and update it.
+    // (Also because we can't clone an TokenserverContext as hawk doesn't derive clone,
+    // but even if we could, we'd run the risk of a consumer using a stale
+    // clone. No idea if this actually makes sense though.)
+    current_context: RefCell<Option<Rc<TokenserverContext>>>,
+    current_backoff: Cell<SystemTime>,
+}
+
+impl TokenserverClient {
+    pub fn new(base_url: String, access_token: String, key_id: String) -> TokenserverClient {
+        TokenserverClient {
+            base_url,
+            access_token,
+            key_id,
+            current_context: RefCell::new(None),
+            current_backoff: Cell::new(UNIX_EPOCH),
+        }
+    }
+
+    pub fn get_context(&self, request_client: &Client) -> Result<Rc<TokenserverContext>> {
```

Comment:
I started wondering if `get_context` could return `&TokenserverContext` instead of `Rc<TokenserverContext>`, but that seems really involved, there's no way to (AFAICT?) decay a `Ref<T>` into a `&T`, and it wouldn't be safe, anyway, since `RefCell` reimplements borrow checking dynamically.